### PR TITLE
SLVS-1558 Refactor SolutionRoslynAnalyzerManager

### PR DESCRIPTION
**Pull request by gabriela-trutan-sonarsource**
[SLVS-1558](https://sonarsource.atlassian.net/browse/SLVS-1558)

Make sure SolutionRoslynAnalyzerManager subscribes to  IActiveConfigScopeTracker.CurrentConfigurationScopeChanged  and
IActiveSolutionTracker.ActiveSolutionChanged events and updates the analyzers accordingly 

[SLVS-1558]: https://sonarsource.atlassian.net/browse/SLVS-1558?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/Infrastructure.VS/Roslyn/SolutionRoslynAnalyzerManager.cs [hunk 1]
```diff
@@ -115,38 +121,11 @@ public void Dispose()
...
-        }
-        else
-        {
-            currentState = new SolutionStateInfo(solutionName);
-            isSameSolution = false;
-        }
-    }
-
     private async Task<ImmutableArray<AnalyzerFileReference>> ChooseAnalyzersAsync() =>
         ChooseAnalyzers(await connectedModeAnalyzerProvider.GetOrNullAsync());
```
Comment (georgii-borovinskikh-sonarsource) [inline src/Infrastructure.VS/Roslyn/SolutionRoslynAnalyzerManager.cs:130]: We should provide the config scope here

File: src/Infrastructure.VS/Roslyn/IRoslynWorkspaceWrapper.cs [hunk 2]
```diff
@@ -36,20 +37,27 @@ internal interface IRoslynWorkspaceWrapper
...
 
-    internal /* for testing */ RoslynWorkspaceWrapper(Workspace workspace)
+    internal /* for testing */ RoslynWorkspaceWrapper(Workspace workspace, IThreadHandling threadHandling)
     {
         this.workspace = workspace;
+        this.threadHandling = threadHandling;
     }
 
     public IRoslynSolutionWrapper CurrentSolution => 
         new RoslynSolutionWrapper(workspace.CurrentSolution);
```
Comment (georgii-borovinskikh-sonarsource) [inline src/Infrastructure.VS/Roslyn/IRoslynWorkspaceWrapper.cs:55]: Can .CurrentSolution be accessed on a background thread?

Comment (gabriela-trutan-sonarsource) [reply src/Infrastructure.VS/Roslyn/IRoslynWorkspaceWrapper.cs:55]: Yes. The only problem was when the TryApplyChanges is called on a background thread

File: src/Infrastructure.VS/Roslyn/SolutionRoslynAnalyzerManager.cs (hunk 1 above)
Comment (gabriela-trutan-sonarsource) [reply src/Infrastructure.VS/Roslyn/SolutionRoslynAnalyzerManager.cs:130]: I don't want to change the interface IconnectedModeAnalyzerProvider in this PR. I added a comment in this task: https://sonarsource.atlassian.net/browse/SLVS-1518

File: src/Infrastructure.VS/Roslyn/SolutionRoslynAnalyzerManager.cs [hunk 3]
```diff
@@ -62,42 +72,38 @@ public SolutionRoslynAnalyzerManager(
...
+        this.asyncLock = asyncLockFactory.Create();
         this.logger = logger;
 
-        connectedModeAnalyzerProvider.AnalyzerUpdatedForConnection += HandleConnectedModeAnalyzerUpdate;
+        activeConfigScopeTracker.CurrentConfigurationScopeChanged += OnConfigurationScopeChanged;
+        activeSolutionTracker.ActiveSolutionChanged += OnActiveSolutionChanged;
     }
 
-    public async Task OnSolutionStateChangedAsync(string solutionName)
+    internal async Task OnSolutionStateChangedAsync(string solutionName)
```
[the remark this answers is not in the document]
Comment (gabriela-trutan-sonarsource) [reply src/Infrastructure.VS/Roslyn/SolutionRoslynAnalyzerManager.cs:93]: I know, we have a ticket in the weekly about it. :D 
I will add it here, though, for consistency.

File: src/Infrastructure.VS/Roslyn/SolutionRoslynAnalyzerManager.cs [hunk 4]
```diff
@@ -160,42 +139,45 @@ private void UpdateAnalyzersIfChanged(ImmutableArray<AnalyzerFileReference> anal
...
+    private void RemoveCurrentAnalyzers()
     {
-        if (!roslynWorkspace.TryApplyChanges(roslynWorkspace.CurrentSolution.RemoveAnalyzerReferences(currentAnalyzers!.Value)))
+        if (currentAnalyzers.HasValue && !roslynWorkspace.TryApplyChanges(roslynWorkspace.CurrentSolution.RemoveAnalyzerReferences(currentAnalyzers.Value)))
         {
-            const string message = "Failed to remove analyzer references while updating analyzers";
-            Debug.Assert(true, message);
-            logger.LogVerbose(message);
+            logger.LogVerbose(Resources.RoslynAnalyzersNotRemoved);
             throw new NotImplementedException();
```
Comment (georgii-borovinskikh-sonarsource) [inline src/Infrastructure.VS/Roslyn/SolutionRoslynAnalyzerManager.cs:157]: Can you fix this as well?

File: src/Infrastructure.VS/Roslyn/SolutionRoslynAnalyzerManager.cs [hunk 5]
```diff
@@ -160,42 +139,45 @@ private void UpdateAnalyzersIfChanged(ImmutableArray<AnalyzerFileReference> anal
...
 
     private void AddAnalyzer(ImmutableArray<AnalyzerFileReference> analyzerToUse)
     {
         if (!roslynWorkspace.TryApplyChanges(roslynWorkspace.CurrentSolution.AddAnalyzerReferences(analyzerToUse)))
         {
-            const string message = "Failed to add analyzer references while adding analyzers";
-            Debug.Assert(true, message);
-            logger.LogVerbose(message);
+            logger.LogVerbose(Resources.RoslynAnalyzersNotAdded);
             throw new NotImplementedException();
```
Comment (georgii-borovinskikh-sonarsource) [inline src/Infrastructure.VS/Roslyn/SolutionRoslynAnalyzerManager.cs:168]: Same

**Review by georgii-borovinskikh-sonarsource — APPROVED** [review]

File: src/Infrastructure.VS/Roslyn/SolutionRoslynAnalyzerManager.cs (hunk 4 above)
Comment (gabriela-trutan-sonarsource) [reply src/Infrastructure.VS/Roslyn/SolutionRoslynAnalyzerManager.cs:157]: You mean to throw a different exception? Or maybe no exception at all.

**Pull request merged**